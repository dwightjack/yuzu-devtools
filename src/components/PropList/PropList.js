import { wire } from 'hyperhtml';
import Prop from '../Prop/Prop';
import PropWatcher from '../PropWatcher/PropWatcher';
import { noop } from '../utils';
import * as styles from './PropList.styles';

const empty = (keys) => wire(keys)`<p class="${styles.empty}">empty object</p>`;

export default function PropList({
  uid,
  title = '',
  onSelect = noop,
  props = {},
  watchers = [],
  watchable = false,
}) {
  const keys = Object.keys(props);
  const list =
    keys.length > 0
      ? keys.map((key) => {
          return Prop({
            key,
            uid,
            watchable,
            value: props[key],
            onSelect,
            watched: watchable && watchers.includes(`${uid}:${key}`),
          });
        })
      : empty(keys);

  const globalWatcher =
    keys.length && watchable
      ? PropWatcher({
          watched: watchers.includes(`${uid}:*`),
          uid,
          onSelect,
          key: '*',
        })
      : '';

  return wire(props)`
    <section class="${styles.root}">
      <h3 class="${styles.title}"><span>${title}</span>${globalWatcher}</h3>
      ${list}
    </section>
  `;
}
