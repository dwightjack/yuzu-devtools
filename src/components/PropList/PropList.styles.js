import { css } from 'emotion';

export const root = css`
  & + & {
    margin-top: var(--gutter);
    border-top: 1px solid var(--color-light);
  }
`;

export const title = css`
  margin: var(--gutter);
  font-size: var(--font-size-m);
  font-weight: bold;
  line-height: 1.2;
`;

export const empty = css`
  margin: 0;
  padding: 0 var(--gutter);
  color: var(--color-quiet);
  font-size: var(--font-size-m);
  font-style: italic;
`;
