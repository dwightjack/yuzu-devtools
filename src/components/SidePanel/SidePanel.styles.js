import { css } from 'emotion';

export const title = css`
  margin: 0;
  padding: 8px var(--gutter);
  font-size: var(--font-size-l);
  font-weight: bold;
  line-height: 1.2;
  background-color: var(--color-lighter);

  & > em {
    font-size: 0.75em;
  }
`;

export const blankSlate = css`
  margin: 0;
  padding: var(--gutter);
  color: var(--color-quiet);
  font-size: var(--font-size-m);
  font-style: italic;
`;
