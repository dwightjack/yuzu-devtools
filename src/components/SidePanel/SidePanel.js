import { wire } from 'hyperhtml';
import PropList from '../PropList/PropList';
import * as styles from './SidePanel.styles';

const blankSlate = wire({ blank: true })`
  <p class="${
    styles.blankSlate
  }">Select a component on the left panel to inspect its properties</p>
`;

export default function SidePanel(props = {}) {
  const { uid, Component, state, options } = props;
  if (!uid) {
    return blankSlate;
  }
  return wire(props)`<section>
    <h2 class="${styles.title}">${Component || 'Component'}<em>#${uid}</em></h2>
    ${PropList({ title: 'Options', props: options })}
    ${PropList({ title: 'State', props: state })}
  </section>`;
}
