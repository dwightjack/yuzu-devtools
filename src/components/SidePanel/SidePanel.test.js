import SidePanel from './SidePanel';
import PropList from '../PropList/PropList';

jest.mock('../PropList/PropList', () => {
  const { mockComponent } = require('../test-utils');
  return jest.fn(mockComponent('PropList'));
});

describe('SidePanel', () => {
  const props = { uid: '_0', Component: 'DemoCmp' };

  test('matches default snapshot', () => {
    expect(SidePanel(props)).toMatchSnapshot();
  });

  test('two "PropList"s', () => {
    PropList.mockClear();
    SidePanel(props);
    expect(PropList).toHaveBeenCalledTimes(2);
  });
});
