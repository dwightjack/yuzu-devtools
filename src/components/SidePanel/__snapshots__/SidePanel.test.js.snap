// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidePanel matches default snapshot 1`] = `
<section
  class="css-lfmitb"
  style=""
>
  
    
  <h2
    class="css-br4xm3"
    style=""
  >
    
    
    DemoCmp
    
    <em
      style=""
    >
      #
      _0
      
    </em>
    
  
  </h2>
  
  
    
  <div
    class="css-10c7rqy"
    style=""
  >
    
      
    <div
      class="css-1ypao5f"
      style=""
    >
      
        
      <proplist
        props={undefined}
        style=""
        title="Options"
      />
      <proplist
        onSelect={[Function]}
        props={undefined}
        style=""
        title="State"
        uid="_0"
        watchable={true}
        watchers={undefined}
      />
      
      
      
    </div>
    
    
  </div>
  
  
</section>
`;
